server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -228,13 +231,13 @@ protected Set<String> createAndUpdateBatches(RoutingAllocation allocation, boole
             batchEntry.getValue().getBatchedShards().forEach(shardId -> currentBatchedShards.put(shardId, batchEntry.getKey()));
         }
 
-        Set<ShardRouting> newShardsToBatch = Sets.newHashSet();
+        Map<ShardId, ShardRouting> newShardsToBatch = new HashMap<>();
         Set<ShardId> batchedShardsToAssign = Sets.newHashSet();
         // add all unassigned shards to the batch if they are not already in a batch
         unassigned.forEach(shardRouting -> {
             if ((currentBatchedShards.containsKey(shardRouting.shardId()) == false) && (shardRouting.primary() == primary)) {
                 assert shardRouting.unassigned();
-                newShardsToBatch.add(shardRouting);
+                newShardsToBatch.put(shardRouting.shardId(), shardRouting);
```

Comment:
As we are overriding previous `ShardRouting` value for an existing `ShardId`, I am assuming we don't really need the contents of `ShardRouting` here. In that case, why don't we just use `Set<ShardId>`

Reply:
`ShardRouting` is used to create `ShardEntry` in line #275. The ShardId is not present in the cache, and we will add a new entry to `ShardsBatch` (after de-duping based on ShardId). If it is an existing ShardId it is added to `batchesToBeAssigned` and does not have entry in ` newShardsToBatch`.

Reply:
you can always do Set.contains?

Reply:
We require`ShardRouting` to assign it for batches (later in line#275) and `ShardId` to de-dupe if the `unassigned` has duplicate entries. Hence using the Map<ShardId,ShardRouting>.